src/hera/workflows/script.py

```diff
@@ -497,6 +497,15 @@ class will be used as inputs, rather than the class itself.
             else:
                 default = MISSING
 
+            type_ = get_origin(func_param.annotation)
+            args = get_args(func_param.annotation)
+            if type_ is Annotated:
+                type_ = get_origin(args[0])
+                args = get_args(args[0])
+
+            if (type_ is Union and len(args) == 2 and type(None) in args) and default is MISSING:
+                raise ValueError(f"Optional parameter '{func_param.name}' doesn't have default value.")
```

Comment:
Small error message suggestion. I would also check that the `default is None`, otherwise it doesn't make sense for it to be Optional
```suggestion
                raise ValueError(f"Optional parameter '{func_param.name}' must have a default value of None.")
```

Reply:
:+1: Fixed!